File: testcases/c/tiny_regex_test.c
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include "re.h"  // This header should declare re_match, re_matchp, re_compile, re_print, etc.

// Helper: a macro to print the result of a test.
#define TEST_MSG(msg)  printf("%s passed.\n", msg)

// Test 1: Literal match
void test_literal_match() {
    int mlen = 0;
    // "hello" should match at offset 0 in "hello world" and match length 5.
    int ret = re_match("hello", "hello world", &mlen);
    assert(ret == 0);
    assert(mlen == 5);
    TEST_MSG("test_literal_match");
}

// Test 2: Literal no-match
void test_literal_no_match() {
    int mlen = 0;
    // "hello" is not found in "world"
    int ret = re_match("hello", "world", &mlen);
    assert(ret == -1);
    TEST_MSG("test_literal_no_match");
}

// Test 3: Beginning anchor
void test_anchor_begin_success() {
    int mlen = 0;
    // "^hello" must match only if string starts with hello.
    int ret = re_match("^hello", "hello world", &mlen);
    assert(ret == 0);
    assert(mlen == 5);
    TEST_MSG("test_anchor_begin_success");
}

void test_anchor_begin_failure() {
    int mlen = 0;
    // Should fail because "hello" is not at the beginning.
    int ret = re_match("^hello", "say hello", &mlen);
    assert(ret == -1);
    TEST_MSG("test_anchor_begin_failure");
}

// Test 4: End anchor
void test_anchor_end_success() {
    int mlen = 0;
    // "world$" should match if the string ends with "world"
    int ret = re_match("world$", "hello world", &mlen);
    // Since the pattern is not anchored at the beginning, the matcher
    // will return the offset of the first character where a match occurs.
    // Here "world" begins at offset 6.
    assert(ret == 6);
    assert(mlen == 5);
    TEST_MSG("test_anchor_end_success");
}

void test_anchor_end_failure() {
    int mlen = 0;
    // "world$" should not match "world hello"
    int ret = re_match("world$", "world hello", &mlen);
    assert(ret == -1);
    TEST_MSG("test_anchor_end_failure");
}

// Test 5: Dot (.) meta-character
void test_dot() {
    int mlen = 0;
    // "h.llo" should match "hello"
    int ret = re_match("h.llo", "hello", &mlen);
    assert(ret == 0);
    assert(mlen == 5);
    TEST_MSG("test_dot");
}

// Test 6: STAR quantifier
void test_star() {
    int mlen = 0;
    // "a*" should match at offset 0 in "aaa". Since '*' can match zero or more,
    // the greedy match here should match all consecutive 'a's.
    int ret = re_match("a*", "aaa", &mlen);
    assert(ret == 0);
    // Depending on implementation details, it should match the entire "aaa"
    // (i.e. mlen == 3) rather than a zero-length match.
    assert(mlen == 3);
    TEST_MSG("test_star");
}

// Test 7: PLUS quantifier
void test_plus() {
    int mlen = 0;
    // "a+" requires at least one 'a'. In "baaab", the first occurrence is at offset 1.
    int ret = re_match("a+", "baaab", &mlen);
    assert(ret == 1);
    assert(mlen == 3);
    TEST_MSG("test_plus");
}

// Test 8: QUESTION quantifier
void test_question() {
    int mlen = 0;
    // "a?" means zero or one 'a'. In a string without 'a', it may match a zero-length string.
    int ret = re_match("a?", "bbb", &mlen);
    // For non-anchored patterns, a zero-length match at offset 0 is acceptable.
    assert(ret == 0);
    // Our implementation may return matchlength 0.
    TEST_MSG("test_question");
}

// Test 9: Character class (simple)
void test_char_class() {
    int mlen = 0;
    // "[abc]+" should match a sequence of a, b, or c.
    // In "xabcx", it should match "abc" starting at offset 1.
    int ret = re_match("[abc]+", "xabcx", &mlen);
    assert(ret == 1);
    assert(mlen == 3);
    TEST_MSG("test_char_class");
}

// Test 10: Range in character class
void test_range() {
    int mlen = 0;
    // "[a-z]+" should match lowercase letters.
    // "HELLO" (all uppercase) should fail.
    int ret = re_match("[a-z]+", "HELLO", &mlen);
    assert(ret == -1);
    TEST_MSG("test_range");
}

// Test 11: Escaped digit (\d)
void test_escaped_digit() {
    int mlen = 0;
    // "\\d+" should match one or more digits.
    // In "abc123xyz", the digits "123" start at offset 3.
    int ret = re_match("\\d+", "abc123xyz", &mlen);
    assert(ret == 3);
    assert(mlen == 3);
    TEST_MSG("test_escaped_digit");
}

// Test 12: Escaped word (\w)
void test_escaped_word() {
    int mlen = 0;
    // "\\w+" should match alphanumeric plus underscore.
    // In "!!!", there are no word characters, so no match.
    int ret = re_match("\\w+", "!!!", &mlen);
    assert(ret == -1);
    TEST_MSG("test_escaped_word");
}

// Test 13: Escaped whitespace (\s)
void test_escaped_whitespace() {
    int mlen = 0;
    // "\\s+" should match one or more whitespace characters.
    // In "abc def", there is a space at offset 3.
    int ret = re_match("\\s+", "abc def", &mlen);
    assert(ret == 3);
    // The match should be the space (" ") so matchlength should be 1.
    assert(mlen == 1);
    TEST_MSG("test_escaped_whitespace");
}

// Test 14: Invalid pattern (unmatched '[')
void test_invalid_pattern() {
    // re_compile should return 0 for an invalid pattern.
    re_t pat = re_compile("[abc");
    assert(pat == 0);
    TEST_MSG("test_invalid_pattern");
}

int main(void) {
    test_literal_match();
    test_literal_no_match();
    test_anchor_begin_success();
    test_anchor_begin_failure();
    test_anchor_end_success();
    test_anchor_end_failure();
    test_dot();
    test_star();
    test_plus();
    test_question();
    test_char_class();
    test_range();
    test_escaped_digit();
    test_escaped_word();
    test_escaped_whitespace();
    test_invalid_pattern();
    printf("All mini regex tests passed.\n");
    return 0;
}

File: testcases/rust/tiny_regex_test.rs
// Assume the following in your lib.rs or module:
// pub type RePattern = Vec<RegexItem>;  // or similar
// pub fn re_compile(pattern: &str) -> Option<RePattern> { ... }
// pub fn re_match(pattern: &str, text: &str) -> Option<(usize, usize)> { ... }

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_literal_match() {
        // "hello" should match at offset 0 in "hello world", with match length 5.
        let result = re_match("hello", "hello world");
        assert_eq!(result, Some((0, 5)));
    }

    #[test]
    fn test_literal_no_match() {
        let result = re_match("hello", "world");
        assert_eq!(result, None);
    }

    #[test]
    fn test_anchor_begin_success() {
        // "^hello" should only match if the text begins with "hello".
        let result = re_match("^hello", "hello world");
        assert_eq!(result, Some((0, 5)));
    }

    #[test]
    fn test_anchor_begin_failure() {
        let result = re_match("^hello", "say hello");
        assert_eq!(result, None);
    }

    #[test]
    fn test_anchor_end_success() {
        // "world$" should match "hello world" starting at offset 6.
        let result = re_match("world$", "hello world");
        assert_eq!(result, Some((6, 5)));
    }

    #[test]
    fn test_anchor_end_failure() {
        let result = re_match("world$", "world hello");
        assert_eq!(result, None);
    }

    #[test]
    fn test_dot() {
        // "h.llo" should match "hello"
        let result = re_match("h.llo", "hello");
        assert_eq!(result, Some((0, 5)));
    }

    #[test]
    fn test_star() {
        // "a*" should match one or more "a" in a greedy manner.
        // For "aaa", expect the match length to be 3.
        let result = re_match("a*", "aaa");
        assert_eq!(result, Some((0, 3)));
    }

    #[test]
    fn test_plus() {
        // "a+" should match at least one "a". In "baaab", expect match at offset 1, length 3.
        let result = re_match("a+", "baaab");
        assert_eq!(result, Some((1, 3)));
    }

    #[test]
    fn test_question() {
        // "a?" should match zero or one "a".
        // In "bbb", a zero-length match at offset 0 is acceptable.
        let result = re_match("a?", "bbb");
        // Depending on implementation, you might get a zero-length match.
        if let Some((offset, len)) = result {
            assert_eq!(offset, 0);
            // Allow len to be either 0 or 1.
            assert!(len == 0 || len == 1);
        } else {
            panic!("Expected a match for a?");
        }
    }

    #[test]
    fn test_char_class() {
        // "[abc]+" should match a sequence of a, b, or c.
        // In "xabcx", expect a match starting at offset 1 with length 3.
        let result = re_match("[abc]+", "xabcx");
        assert_eq!(result, Some((1, 3)));
    }

    #[test]
    fn test_range() {
        // "[a-z]+" should match lowercase letters.
        // In "HELLO", expect no match (if case sensitive).
        let result = re_match("[a-z]+", "HELLO");
        assert_eq!(result, None);
    }

    #[test]
    fn test_escaped_digit() {
        // "\\d+" should match one or more digits.
        // In "abc123xyz", expect match at offset 3, length 3.
        let result = re_match("\\d+", "abc123xyz");
        assert_eq!(result, Some((3, 3)));
    }

    #[test]
    fn test_escaped_word() {
        // "\\w+" should match alphanumeric (and underscore).
        // In "!!!", expect no match.
        let result = re_match("\\w+", "!!!");
        assert_eq!(result, None);
    }

    #[test]
    fn test_escaped_whitespace() {
        // "\\s+" should match one or more whitespace characters.
        // In "abc def", expect match at offset 3, length 1.
        let result = re_match("\\s+", "abc def");
        assert_eq!(result, Some((3, 1)));
    }

    #[test]
    fn test_invalid_pattern() {
        // An invalid pattern (e.g. unmatched '[') should result in None.
        let pat = re_compile("[abc");
        assert!(pat.is_none());
    }
}
